client/xdag_config.c

```diff
@@ -0,0 +1,536 @@
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+//#include <errno.h>
+#include "xdag_config.h"
+
+
+
+#ifndef true
+#define true (1)
+#endif
+
+#ifndef false
+#define false (0)
+#endif
+
+#if defined (WIN32)
+#define strkscmp stricmp
+#else
+#define strkscmp strcasecmp
+#endif
+
+static char *trim(char *str)
+{
+    if(str != NULL)
+    {
+        int len = (int)strlen(str);
+        char *start = str;
+        char *end = str + len;
+        while(start < end && *start <= ' ')
+        {
+            start++;
+        }
+        while(end > start && *(end-1) <= ' ')
+        {
+            end--;
+        }
+        if(end - start != len)
+        {
+            strncpy(str, start, end-start);
+            str[end-start] = '\0';
+        }
+    }
+    return str;
+}
+
+
+
+
+
+typedef struct
+{
+    char *name;
+    char *value;
+} KEY;
+
+typedef struct
+{
+    char *name;
+    int number_keys;
+    KEY **keys;
+    char **key_names;
+} NODE;
+
+
+
+typedef struct
+{
+    char *path;
+    int number_nodes;
+    NODE **nodes;
+    char **node_names;
+} CFG;
+
+
+static const char COMMENT_MARK[] = {'#', '*'};
+typedef enum
+{
+    TYPE_EMPTYLINE,
+	TYPE_COMMENT,
+    TYPE_NODE,
+    TYPE_KEY,
+    TYPE_UNKNOW,
+} CFG_CONTENT_TYPE;
+
+static char *readline(FILE *fp)
+{
+    char buf[128];
+    char c;
+    char *line = NULL;
+    int line_len = 0, buf_len = 0;
+
+    while(!feof(fp))
+    {
+        memset(buf, 0, sizeof(buf));
+        fgets(buf, sizeof(buf), fp);
+        buf_len = strlen(buf);
+        line = (char *)realloc(line, line_len+buf_len+1);
+        strcpy(line+line_len, buf);
+        line_len += buf_len;
+
+        c = buf[sizeof(buf)-2];
```

Comment:
Why `buf[sizeof(buf)-2]`?

Reply:
The allocated buf is a pointer array of 128, sizeof(buf)=128, minus 2 is to get the last character of the array.

Reply:
Thank you, I am going to modify it.